Automattic/jetpack
# WAF: Connect the UI to actually updating the IP block and allow lists when saving the settings.

This PR connects the UI that is introduced in #23769 to the actual functionality. This means that when the user updates the lists of IPs to be blocked or allowed, they actually have that effect now.

Technically it will probably not matter which one of the two PRs (the linked one above or this one) will be merged first, but for simplicity we should probably merge the linked one first.





#### Changes proposed in this Pull Request:

* Hooked into the option updates to make changes to the allow-ip.php and block-ip.php.
* Removed some code that was handling empty option values, as we always get an empty array now even if the option does not yet exist.

#### Jetpack product discussion



1201069996155224-as-1202148145416678

#### Does this pull request change what data or activity we track or use?



No. 

#### Testing instructions:






* Enable the Settings UI (see #23769 for details on how to do that)
* Check out this branch.
* Remove any prior `jetpack_waf_ip_allow_list` and `jetpack_waf_ip_block_list` options if you still have them in your database. (the format changed since prior PRs, so you may have unexpected test data in there)
* Save both valid and invalid IP addresses in the allow list and block list settings - the IPs are split on whitespace (including line breaks), commas or semicolons. Be creative. Example value:
```
8.8.8.8
4.4.4.4,2.2.2.2;1.1.1.1 9.9.9.9,5555.5555.5555.5555
```
* Verify that only the valid IPs (and all of them) end up in the correct PHP files when you save the settings. From inside the WordPress docker container, you can do `cat /usr/local/src/jetpack-monorepo/projects/packages/waf/rules/block-ip.php` (or `allow-ip.php` respectively) to do that.
* Add your own IP to the allow-list -> check your requests are always allowed (verifiable by adding `?jetpack-waf-always-block=1` query parameter) 
* **_BEFORE_** doing the next steps: Be sure you can access your database and CLI because all of your requests will be blocked otherwise.
* Remove your own IP from allow-list and add it to block-list -> save settings and verify all your requests are blocked (including wp-admin, meaning you can also not change the settings anymore).

If you want to unblock yourself again, you can do that by: 
* Editing the block-list option `jetpack_waf_ip_block_list` manually in the database and remove your IP.
* Running `wp jetpack-waf setup normal --user=1 --allow-root` to re-generate the files.

Alternatively, you can edit your `block-ip.php` file manually of course, but this is not how it should be used.


##### Additional test scenario for the "on/off" toggle
This part of the code removes all IPs from the `block-ip.php` and `allow-ip.php` files, effectively disabling the feature.

* Add your IP to the allow-list.
* Verify you can make a request to the callback despite adding the `?jetpack-waf-always-block=1` parameter. 
* Toggle the list-feature off
* Verify you can not make the same request anymore (because you are not on the allow-list anymore)
* Toggle the list-feature on again
* Verify your request is allowed again
